acb_dirichlet/zeta_zero.c

```diff
@@ -0,0 +1,748 @@
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <http://www.gnu.org/licenses/>.
+*/
+
+#include "acb_dirichlet.h"
+#include "arb_calc.h"
+
+static void
+_acb_set_arf(acb_t res, const arf_t t)
+{
+    acb_zero(res);
+    arb_set_arf(acb_realref(res), t);
+}
+
+static int
+_definite_hardy_z(arb_t res, const arf_t t, slong *pprec)
+{
+    int msign;
+    acb_t z;
+    acb_init(z);
+    while (1)
+    {
+        _acb_set_arf(z, t);
+        acb_dirichlet_hardy_z(z, z, NULL, NULL, 1, *pprec);
+        msign = arb_sgn_nonzero(acb_realref(z));
+        if (msign)
+        {
+            break;
+        }
+        *pprec *= 2;
+    }
+    acb_get_real(res, z);
+    acb_clear(z);
+    return msign;
+}
+
+typedef struct _zz_node_struct
+{
+    arf_struct t;
+    arb_struct v;
+    fmpz *gram;
+    slong prec;
+    struct _zz_node_struct *prev;
+    struct _zz_node_struct *next;
+}
+zz_node_struct;
+
+typedef zz_node_struct zz_node_t[1];
+typedef zz_node_struct * zz_node_ptr;
+typedef const zz_node_struct * zz_node_srcptr;
+
+static int
+zz_node_is_gram_node(const zz_node_t p)
+{
+    return p->gram != NULL;
+}
+
+static int
+zz_node_sgn(const zz_node_t p)
+{
+    int s = arb_sgn_nonzero(&p->v);
+    if (!s)
+    {
+        flint_printf("unexpectedly imprecise evaluation of Z(t)\n");
+        flint_abort();
+    }
+    return s;
+}
+
+static int
+zz_node_is_good_gram_node(const zz_node_t p)
+{
+    if (zz_node_is_gram_node(p))
+    {
+        int s = zz_node_sgn(p);
+        if ((s > 0 && fmpz_is_even(p->gram)) ||
+            (s < 0 && fmpz_is_odd(p->gram)))
+        {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+static void
+zz_node_init(zz_node_t p)
+{
+    arf_init(&p->t);
+    arb_init(&p->v);
+    arb_indeterminate(&p->v);
+    p->prec = 0;
+    p->gram = NULL;
+    p->prev = NULL;
+    p->next = NULL;
+}
+
+static void
+zz_node_clear(zz_node_t p)
+{
+    arf_clear(&p->t);
+    arb_clear(&p->v);
+    if (p->gram)
+    {
+        fmpz_clear(p->gram);
+        flint_free(p->gram);
+    }
+    p->prec = 0;
+    p->gram = NULL;
+    p->prev = NULL;
+    p->next = NULL;
+}
+
+static int
+zz_node_refine(zz_node_t p)
+{
+    p->prec = 2*FLINT_MAX(p->prec, 8);
+    return _definite_hardy_z(&p->v, &p->t, &p->prec);
+}
+
+static void
+refine_one(zz_node_t a, zz_node_t b)
+{
+    zz_node_refine((a->prec < b->prec) ? a : b);
+}
+
+static zz_node_ptr
+create_non_gram_node(const arf_t t)
+{
+    zz_node_ptr p = flint_malloc(sizeof(zz_node_struct));
+    zz_node_init(p);
+    arf_set(&p->t, t);
+    p->prec = 8;
+    _definite_hardy_z(&p->v, &p->t, &p->prec);
+    return p;
+}
+
+static zz_node_ptr
+create_gram_node(const fmpz_t n)
+{
+    zz_node_ptr p;
+    arb_t t, v;
+    acb_t z;
+    slong prec = 8;
+
+    arb_init(t);
+    arb_init(v);
+    acb_init(z);
+
+    arb_indeterminate(v);
+    while (arb_contains_zero(v))
+    {
+        prec *= 2;
```

Comment:
I want to move this to the end of the loop, for consistency with _definite_hardy_z.